File: frontend/src/types/exam.ts
// Exam types
export interface ExamQuestion {
  id: number;
  question_id: number;
  question_order: number;
  choice_order: string;
  exam_id: number;
  created_at: string;
}

export interface Exam {
  id: number;
  code: string;
  title: string;
  subject: string;
  duration: number; // in minutes
  total_questions: number;
  description?: string;
  is_active: boolean;
  created_by: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string | null;
}

export interface ExamWithQuestions extends Exam {
  exam_questions: ExamQuestion[];
}

export interface ExamCreate {
  code: string;
  title: string;
  subject: string;
  duration: number;
  total_questions: number;
  description?: string;
  is_active?: boolean;
}

export interface ExamUpdate {
  code?: string;
  title?: string;
  subject?: string;
  duration?: number;
  total_questions?: number;
  description?: string;
  is_active?: boolean;
}

export interface ExamGenerateRequest {
  code: string;
  title: string;
  subject: string;
  duration: number;
  total_questions: number;
  description?: string;
  shuffle_choices?: boolean;
}

export interface ExamQuestionDetail {
  id: number;
  question_id: number;
  question_order: number;
  content: string;
  content_img?: string;
  choices: string[]; // Shuffled choices
  choice_labels: string[]; // Corresponding labels (A, B, C, D)
  correct_answer_index: number; // Index of correct answer in shuffled choices
  mark: number;
  unit?: string;
}

export interface ExamDetailResponse extends Exam {
  questions: ExamQuestionDetail[];
  creator_username?: string;
}

export interface ExamListParams {
  page?: number;
  page_size?: number;
  subject?: string;
  created_by?: number;
}

export interface ExamStats {
  total_exams: number;
  active_exams: number;
  subjects: string[];
  avg_duration: number;
  avg_questions: number;
}

File: backend/app/schemas/exam.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, field_validator


class ExamQuestionBase(BaseModel):
    question_id: int
    question_order: int
    choice_order: str


class ExamQuestionCreate(ExamQuestionBase):
    pass


class ExamQuestionOut(ExamQuestionBase):
    id: int
    exam_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class ExamBase(BaseModel):
    code: str
    title: str
    subject: str
    duration: int  # in minutes
    total_questions: int
    description: Optional[str] = None
    is_active: bool = True

    @field_validator("duration")
    @classmethod
    def validate_duration(cls, v):
        if v <= 0:
            raise ValueError("Duration must be greater than 0")
        return v

    @field_validator("total_questions")
    @classmethod
    def validate_total_questions(cls, v):
        if v <= 0:
            raise ValueError("Total questions must be greater than 0")
        return v


class ExamCreate(ExamBase):
    pass


class ExamUpdate(BaseModel):
    code: Optional[str] = None
    title: Optional[str] = None
    subject: Optional[str] = None
    duration: Optional[int] = None
    total_questions: Optional[int] = None
    description: Optional[str] = None
    is_active: Optional[bool] = None

    @field_validator("duration")
    @classmethod
    def validate_duration(cls, v):
        if v is not None and v <= 0:
            raise ValueError("Duration must be greater than 0")
        return v

    @field_validator("total_questions")
    @classmethod
    def validate_total_questions(cls, v):
        if v is not None and v <= 0:
            raise ValueError("Total questions must be greater than 0")
        return v


class ExamOut(ExamBase):
    id: int
    created_by: int
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ExamWithQuestions(ExamOut):
    exam_questions: List[ExamQuestionOut] = []

    class Config:
        from_attributes = True


class ExamGenerateRequest(BaseModel):
    """Request schema for generating a new exam"""
    code: str
    title: str
    subject: str
    duration: int  # in minutes
    total_questions: int
    description: Optional[str] = None
    shuffle_choices: bool = True  # Whether to shuffle answer choices

    @field_validator("duration")
    @classmethod
    def validate_duration(cls, v):
        if v <= 0:
            raise ValueError("Duration must be greater than 0")
        return v

    @field_validator("total_questions")
    @classmethod
    def validate_total_questions(cls, v):
        if v <= 0:
            raise ValueError("Total questions must be greater than 0")
        return v


class ExamQuestionDetail(BaseModel):
    """Detailed question info for exam display"""
    id: int
    question_id: int
    question_order: int
    content: str
    content_img: Optional[str] = None
    choices: List[str]  # Shuffled choices
    choice_labels: List[str]  # Corresponding labels (A, B, C, D)
    correct_answer_index: int  # Index of correct answer in shuffled choices
    mark: float
    unit: Optional[str] = None

    class Config:
        from_attributes = True


class ExamDetailResponse(ExamOut):
    """Detailed exam response with questions"""
    questions: List[ExamQuestionDetail] = []
    creator_username: Optional[str] = None

    class Config:
        from_attributes = True
